src/webgpu/api/operation/vertex_state/index_format.spec.ts

```diff
@@ -0,0 +1,313 @@
+import { GPUTest } from '../../../gpu_test.js';
+import { getTextureCopyLayout } from '../../../util/texture/layout.js';
+
+const kHeight = 4;
+const kWidth = 4;
+const kTextureFormat = 'bgra8unorm' as const;
+const kGreen = new Uint8Array([0x00, 0xff, 0x00, 0xff]); // green
+const kBlack = new Uint8Array([0x00, 0x00, 0x00, 0xff]); // black
+
+type Raster4x4 = readonly [
+  readonly [0 | 1, 0 | 1, 0 | 1, 0 | 1],
+  readonly [0 | 1, 0 | 1, 0 | 1, 0 | 1],
+  readonly [0 | 1, 0 | 1, 0 | 1, 0 | 1],
+  readonly [0 | 1, 0 | 1, 0 | 1, 0 | 1]
+];
+
+/** Expected 4x4 rasterization of 4 points. */
+const kPoints: Raster4x4 = [
+  [1, 0, 0, 1],
+  [0, 0, 0, 0],
+  [0, 0, 0, 0],
+  [1, 0, 0, 1],
+];
+
+/** Expected 4x4 rasterization of a diagonal line. */
+const kLine: Raster4x4 = [
+  [1, 0, 0, 0],
+  [0, 1, 0, 0],
+  [0, 0, 1, 0],
+  [0, 0, 0, 1],
+];
+
+/** Expected 4x4 rasterization of two lines forming an X. */
+const kXShape: Raster4x4 = [
+  [1, 0, 0, 1],
+  [0, 1, 1, 0],
+  [0, 1, 1, 0],
+  [1, 0, 0, 1],
+];
+
+/** Expected 4x4 rasterization of a bottom-left triangle. */
+const kBottomLeftTriangle: Raster4x4 = [
+  [0, 0, 0, 0],
+  [1, 0, 0, 0],
+  [1, 1, 0, 0],
+  [1, 1, 1, 0],
+];
+
+/** Expected 4x4 rasterization of two overlapping triangles in a concave shape. */
+const kConcaveShape: Raster4x4 = [
+  [0, 0, 0, 1],
+  [1, 0, 1, 1],
+  [1, 1, 1, 1],
+  [1, 1, 1, 1],
+];
+
+/** Expected 4x4 rasterization filling the whole quad. */
+const kSquare: Raster4x4 = [
+  [1, 1, 1, 1],
+  [1, 1, 1, 1],
+  [1, 1, 1, 1],
+  [1, 1, 1, 1],
+];
+
+/** Expected 4x4 rasterization with no pixels. */
+const kNothing: Raster4x4 = [
+  [0, 0, 0, 0],
+  [0, 0, 0, 0],
+  [0, 0, 0, 0],
+  [0, 0, 0, 0],
+];
+
+const kPrimitiveTopologiesForRestart = [
+  { primitiveTopology: 'point-list', _expectedShape: kPoints },
+  { primitiveTopology: 'line-list', _expectedShape: kLine },
+  { primitiveTopology: 'line-strip', _expectedShape: kXShape },
+  { primitiveTopology: 'triangle-list', _expectedShape: kBottomLeftTriangle },
+  { primitiveTopology: 'triangle-strip', _expectedShape: kConcaveShape },
+] as const;
+
+const { byteLength, bytesPerRow, rowsPerImage } = getTextureCopyLayout(kTextureFormat, '2d', [
+  kWidth,
+  kHeight,
+  1,
+]);
+
+class IndexFormatTest extends GPUTest {
+  MakeRenderPipeline(
+    primitiveTopology: GPUPrimitiveTopology,
+    indexFormat?: GPUIndexFormat
+  ): GPURenderPipeline {
+    const vertexModule = this.device.createShaderModule({
+      code: `
+        const pos: array<vec2<f32>, 4> = array<vec2<f32>, 4>(
+          vec2<f32>(-1.0, 1.0),
+          vec2<f32>(1.0, -1.0),
+          vec2<f32>(1.0, 1.0),
+          vec2<f32>(-1.0, -1.0));
+
+        [[builtin(position)]] var<out> Position : vec4<f32>;
+        [[builtin(vertex_idx)]] var<in> VertexIndex : u32;
+
+        [[stage(vertex)]]
+        fn main() -> void {
+          Position = vec4<f32>(pos[VertexIndex], 0.0 , 1.0);
```

Comment:
Handle `VertexIndex == 0xFFFF || VertexIndex == 0xFFFFFFFF` explicitly here. OK to just do both because we don't use 0xFFFF in any of the uint32 index buffers. (Though maybe we should have a test for that too)